src/main/kotlin/org/rust/lang/core/completion/RsMacroCompletionProvider.kt

```diff
@@ -9,34 +9,63 @@ import com.intellij.codeInsight.completion.CompletionParameters
 import com.intellij.codeInsight.completion.CompletionProvider
 import com.intellij.codeInsight.completion.CompletionResultSet
 import com.intellij.patterns.ElementPattern
-import com.intellij.patterns.PatternCondition
 import com.intellij.patterns.PlatformPatterns.psiElement
+import com.intellij.psi.PsiComment
 import com.intellij.psi.PsiElement
+import com.intellij.psi.PsiErrorElement
+import com.intellij.psi.PsiWhiteSpace
 import com.intellij.util.ProcessingContext
 import org.rust.lang.RsLanguage
 import org.rust.lang.core.RsPsiPattern
+import org.rust.lang.core.psi.RsElementTypes.COLONCOLON
 import org.rust.lang.core.psi.RsElementTypes.IDENTIFIER
 import org.rust.lang.core.psi.RsMacro
-import org.rust.lang.core.psi.RsMacroCall
-import org.rust.lang.core.psi.RsPath
-import org.rust.lang.core.psi.RsPathExpr
-import org.rust.lang.core.psi.ext.RsItemElement
-import org.rust.lang.core.psi.ext.RsMod
-import org.rust.lang.core.psi.ext.ancestorStrict
-import org.rust.lang.core.psi.ext.queryAttributes
+import org.rust.lang.core.psi.ext.*
+import org.rust.lang.core.psi.unescapedText
 import org.rust.lang.core.psiElement
+import org.rust.lang.core.resolve.ScopeEntry
 import org.rust.lang.core.resolve.collectCompletionVariants
-import org.rust.lang.core.resolve.processMacroCallVariants
+import org.rust.lang.core.resolve.processMacroCallVariantsInScope
+import org.rust.lang.core.resolve.processMacrosExportedByCrateName
 import org.rust.lang.core.withPrevSiblingSkipping
 
+/**
+ * This completion provider is used only in the case of incomplete macro call outside of function or other code block.
+ * ```
+ * thread_lo/*caret*/
+ * std::thread_lo/*caret*/
+ * fn main() {}
+ * ```
+ */
 object RsMacroCompletionProvider : CompletionProvider<CompletionParameters>() {
-    override fun addCompletions(parameters: CompletionParameters, context: ProcessingContext, result: CompletionResultSet) {
-        val mod = parameters.position.ancestorStrict<RsMod>()
-        result.addAllElements(collectCompletionVariants { variantsCollector ->
-            processMacroCallVariants(parameters.position) { entry ->
+    override fun addCompletions(
+        parameters: CompletionParameters,
+        context: ProcessingContext,
+        result: CompletionResultSet
+    ) {
+        val position = parameters.position
+        val rsElement = position.ancestorStrict<RsElement>() ?: return
+        val mod = rsElement.ancestorOrSelf<RsMod>()
+
+        val leftSiblings = position.leftSiblings
+            .filter { it !is PsiWhiteSpace && it !is PsiComment && it !is PsiErrorElement }
+            .take(3).toList()
+        val is2segmentPath = leftSiblings.getOrNull(0)?.elementType == COLONCOLON &&
+            leftSiblings.getOrNull(1)?.elementType == IDENTIFIER
+            leftSiblings.getOrNull(2)?.elementType != COLONCOLON
```

Comment:
Hm, why do we need this condition?
It seems `::std::thread_local! { }` is valid

Reply:
We want to forbid a::b::c. More complex condition is needed, I guess.